src/Disks/S3/ProxyResolverConfiguration.cpp

```diff
@@ -53,20 +64,40 @@ Aws::Client::ClientConfigurationPerRequest ProxyResolverConfiguration::getConfig
 
         LOG_DEBUG(&Poco::Logger::get("AWSClient"), "Use proxy: {}://{}:{}", proxy_scheme, proxy_host, proxy_port);
 
-        cfg.proxyScheme = Aws::Http::SchemeMapper::FromString(proxy_scheme.c_str());
-        cfg.proxyHost = proxy_host;
-        cfg.proxyPort = proxy_port;
+        cached_config.proxyScheme = Aws::Http::SchemeMapper::FromString(proxy_scheme.c_str());
+        cached_config.proxyHost = proxy_host;
+        cached_config.proxyPort = proxy_port;
+        cache_timestamp = time(nullptr);
 
-        return cfg;
+        return cached_config;
     }
     catch (...)
     {
         tryLogCurrentException("AWSClient", "Failed to obtain proxy");
         /// Don't use proxy if it can't be obtained.
+        Aws::Client::ClientConfigurationPerRequest cfg;
         return cfg;
     }
 }
 
+void ProxyResolverConfiguration::errorReport(const Aws::Client::ClientConfigurationPerRequest & config)
+{
+    if (config.proxyHost.empty())
+        return;
+
+    std::unique_lock lock(cache_mutex);
+
+    if (!cache_ttl || !cache_timestamp || time(nullptr) > cache_timestamp + cache_ttl)
+        return;
+
+    if (cached_config.proxyScheme != config.proxyScheme || cached_config.proxyHost != config.proxyHost
```

Comment:
Is it required not to invalidate proxy if it already updated?

Reply:
Yes, in case when CH make several requests in parallel. Sample timeline:
- request1 get proxy1 from remote resolver and put it in cache
- request2 get proxy1 from cache
- request1 failed, invalidate proxy1 in cache
- request3 get proxy2 from remote resolver and put it in cache
- request2 failed. In cache now proxy2, when failed request with proxy1